scriptmodules/emulators/hypseus.sh

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env bash
+
+# This file is part of The RetroPie Project
+#
+# The RetroPie Project is the legal property of its developers, whose names are
+# too numerous to list here. Please refer to the COPYRIGHT.md file distributed with this source.
+#
+# See the LICENSE.md file at the top-level directory of this distribution and
+# at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
+#
+
+rp_module_id="hypseus"
+rp_module_desc="Hypseus Singe - Laserdisc Emulator"
+rp_module_help="ROM Extension: .daphne\n\nCopy your Daphne roms to $romdir/daphne"
+rp_module_licence="GPL3 https://raw.githubusercontent.com/DirtBagXon/hypseus-singe/master/LICENSE"
+rp_module_repo="git https://github.com/DirtBagXon/hypseus-singe.git RetroPie"
+rp_module_section="exp"
+rp_module_flags="sdl2"
+
+function depends_hypseus() {
+    getDepends libvorbis-dev libogg-dev zlib1g-dev libmpeg2-4-dev libsdl2-image-dev libsdl2-ttf-dev cmake
+}
+
+function sources_hypseus() {
+    gitPullOrClone "$md_build" "$md_repo_url" "$md_repo_branch" "" 0
+}
+
+function build_hypseus() {
+    local params=()
+    mkdir build
+    cd build
+    cmake ../src
+    make
+    cp hypseus ../hypseus.bin
+    md_ret_require="hypseus"
+}
+
+function install_hypseus() {
+    md_ret_files=(
+        'sound'
+        'pics'
+        'fonts'
+        'hypseus.bin'
+        'LICENSE'
+    )
+}
+
+function configure_hypseus() {
+    mkRomDir "daphne"
+    mkRomDir "daphne/roms"
+
+    addEmulator 1 "$md_id" "hypseus" "$md_inst/hypseus.sh %ROM%"
+    addSystem "hypseus"
+
+    [[ "$md_mode" == "remove" ]] && return
+
+    mkUserDir "$md_conf_root/hypseus"
+
+    if [[ ! -f "$md_conf_root/hypseus/hypinput.ini" ]]; then
+        cp -v "$md_data/hypinput.ini" "$md_conf_root/hypseus/hypinput.ini"
+    fi
+    ln -snf "$romdir/daphne/roms" "$md_inst/roms"
+    ln -snf "$romdir/daphne/roms" "$md_inst/singe"
+    ln -sf "$md_conf_root/$md_id/hypinput.ini" "$md_inst/hypinput.ini"
+
+    cat >"$md_inst/hypseus.sh" <<_EOF_
+#!/bin/bash
+dir="\$1"
+name="\${dir##*/}"
+name="\${name%.*}"
+
+if [[ -f "\$dir/\$name.commands" ]]; then
+    params=\$(<"\$dir/\$name.commands")
+fi
+
+if [[ -f "\$dir/\$name.singe" ]]; then
+    "$md_inst/hypseus.bin" singe vldp -retropath -framefile "\$dir/\$name.txt" -script "\$dir/\$name.singe" -homedir "$md_inst" -fullscreen \$params
+else
+    "$md_inst/hypseus.bin" "\$name" vldp -framefile "\$dir/\$name.txt" -homedir "$md_inst" -fullscreen \$params
+fi
+_EOF_
+    chmod +x "$md_inst/hypseus.sh"
+
+    chown -R $user:$user "$md_inst"
```

Comment:
Is regular user _write_ access really needed for the installation folder ?

Reply:
Yes user access is required for screenshots and logs, same as Daphne.

I can be more restrictive in hypseus though, via:

```
    mkUserDir "$md_inst/logs"
    mkUserDir "$md_inst/screenshots"
    chown -R $user:$user "$md_inst/logs"
    chown -R $user:$user "$md_inst/screenshots"
```

If that works ?

Reply:
Also 'ram' folder where certain Daphne games store highscore tables etc.

Looked at this, but can't really see how to limit further without creating user writable directories in some other location...

Seems most obvious to leave in the main `-homedir` folder, like Daphne always has.

Reply:
A solution would be to use the _configdir_ for writable data and point `-homedir` to it. One - more - advantage is being able to copy the config folder between systems and keep the user generated content (e.g. hiscore tables). _configdir_ is already user writeable.

Reply:
It might become a bit of a symlinking hell, but would involve creating a few more dirs (in  _$md_inst_ and _$md_conf_root/daphne/_) on install and symlinking would look something like this:

```
total 1872
drwxr-xr-x 9 pi   pi      4096 Aug 22 13:57 .
drwxr-xr-x 5 root root    4096 Aug 22 13:51 ..
drwxr-xr-x 2 pi   pi      4096 Aug 22 13:51 fonts
drwxr-xr-x 2 pi   pi      4096 Aug 22 13:57 framefile
lrwxrwxrwx 1 pi   pi        41 Aug 22 13:51 hypinput.ini -> /opt/retropie/configs/daphne/hypinput.ini
-rwxr-xr-x 1 pi   pi   1836348 Aug 22 21:14 hypseus.bin
-rwxr-xr-x 1 pi   pi       528 Aug 22 13:51 hypseus.sh
-rw-r--r-- 1 pi   pi     32472 Aug 22 13:51 LICENSE
drwxr-xr-x 2 pi   pi      4096 Aug 22 13:57 logs -> /opt/retropie/configs/daphne/logs
drwxr-xr-x 3 pi   pi      4096 Aug 22 13:51 pics
drwxr-xr-x 2 pi   pi      4096 Aug 22 19:26 ram -> /opt/retropie/configs/daphne/ram
-rw-r--r-- 1 pi   pi       286 Aug 22 13:51 retropie.pkg
lrwxrwxrwx 1 pi   pi        34 Aug 22 13:51 roms -> /home/pi/RetroPie/roms/daphne/roms
drwxr-xr-x 2 pi   pi      4096 Aug 22 13:57 screenshots -> /opt/retropie/configs/daphne/screenshots
lrwxrwxrwx 1 pi   pi        34 Aug 22 13:51 singe -> /home/pi/RetroPie/roms/daphne/roms
drwxr-xr-x 2 pi   pi      4096 Aug 22 13:51 sound
```

I can look at that, if it's something you would want to pursue ?

Reply:
Moving _-homedir_ would create a lot more issues, as Daphne is notoriously fussy about its paths.

I think symlinking, as per above, would be simpler.

Reply:
Well, given the constraints, I guess using `$md_inst` for `homedir` is ok and there's no need to complicate the installation layout in this case.

Reply:
I would still be keen to avoid a writable md_inst. Unless it needs to make random named temporary files
It should be avoidable. Just to note `$md_inst` is wiped on update
So anything user writable needs to be symlinked etc.

Reply:
Another option to consider would be to symlink the other way around. Depending what static data is needed relative to `-homedir` it could be symlinked from the config area to `$md_inst`. Another option would be to adjust the source since this is a fork anyway with changes. But whichever seems most reasonable.

If we really can't avoid a writable md_inst, I may be persuaded but I'd like to avoid it. Thanks :-)

Reply:
Rather than use the _$md_conf_root_ for _non-conf_ data, looking at some other packages would **$home** be appropriate for savegame data, logs and screenshots ?

e.g. from _fs-uae.sh_

    mkUserDir "$home/Documents/FS-UAE"
    mkUserDir "$home/Documents/FS-UAE/Configurations"

_i.e._

    mkUserDir "$home/hypseus/ram"
    mkUserDir "$home/hypseus/logs"
    mkUserDir "$home/hypseus/screenshots"

And then create appropriate symlinks?

That would leave only '_framefile_' which I can create directory on install. Nothing in RetroPie should be writing to that. Looking at source, this is just used as a secondary check for missing framefiles by the VLDP. So has no need to be writable by the user, but will try to created if it doesn't exist.

Reply:
See https://github.com/RetroPie/RetroPie-Setup/pull/3387/commits/cfcf2a1f274cac397e89f964a44e8f80807f31a8